File: 1325.删除给定值的叶子节点/1325.删除给定值的叶子节点.py
'''
Descroption: LeetCode 1325. 删除给定值的叶子节点
Author: EmoryHuang
Date: 2021-08-01 13:07:05
Method:
递归，删除所有值为 target 的叶子节点
'''

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def removeLeafNodes(self, root: TreeNode, target: int) -> TreeNode:
        if not root:
            return None
        root.left = self.removeLeafNodes(root.left, target)
        root.right = self.removeLeafNodes(root.right, target)
        if not root.left and not root.right and root.val == target:
            return None
        return root

File: Solution/1325.删除给定值的叶子节点/1325.删除给定值的叶子节点.cpp
/*
 * @Descroption: LeetCode 1325. 删除给定值的叶子节点
 * @Author: EmoryHuang
 * @Date: 2021-08-01 13:08:49
 * @Method:
 * 递归，删除所有值为 target 的叶子节点
 */

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
   public:
    TreeNode* removeLeafNodes(TreeNode* root, int target) {
        if (!root) return nullptr;
        root->left = removeLeafNodes(root->left, target);
        root->right = removeLeafNodes(root->right, target);
        if (!root->left && !root->right && root->val == target) return nullptr;
        return root;
    }
};